### PR TITLE
Ticket2130 mk2 chopper

### PR DESCRIPTION
**Pull request by AdrianPotter**
### Description of work

Create the IOC for a MK2 chopper

### To test

https://github.com/ISISComputingGroup/IBEX/issues/2130

### Acceptance criteria

Test that the IOC works with the device emulator and OPI to communicate with a MK2 chopper. We don't currently have access to a real one so we have to trust the emulator for the time being.

Emulator: https://github.com/ISISComputingGroup/EPICS-DeviceEmulator/pull/11
OPI: https://github.com/ISISComputingGroup/ibex_gui/pull/479
EPICS: https://github.com/ISISComputingGroup/EPICS/pull/79
Support module: https://github.com/ISISComputingGroup/EPICS-Mk2Chopper

---

#### Code Review

- [ ] Is the code of an acceptable quality?
- [ ] Are the PVs named according to the [naming standards](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/PV-Naming).
- [ ] Have suitable [disable records](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Disable-records) been added?
- [ ] Have suitable [simulation records](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Record-Simulation) been added?
- [ ] Have the [finishing touches](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/IOC-Finishing-Touches) been applied?
- [ ] Is there an [emulator](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Emulating-Devices) for the device?
- [ ] Have the changes been documented in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev). If so, do they describe the changes appropriately?
- [ ] If an OPI has been modified, does it conform to the [style guidelines](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/OPI-Creation)? There is a script called `check_opi_format.py` to help with this.

### Functional Tests

- [ ] Do changes function as described? Add comments below that describe the tests performed.
- [ ] Does the IOC respond correctly both in full and simulation mode, where it's possible to test both?
- [ ] If there are multiple _0n IOCs, do they run correctly?

### Final steps

- [ ] Reviewer has updated the submodule in the main EPICS repo? See **Reviewing work for the subModules of EPICS** in the [Git workflow](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Git-workflow) page for details.
- [ ] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section

### CODE REVIEW

File: MK2CHOPR/MK2CHOPR-IOC-01App/Db/MK2CHOPR.db [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+#######################################################
+# Start/stop
+#######################################################
+
+record(bo, "$(P)ON:SP")
+{
+    field(DESC, "Set the chopper off(0)/on(1)")
+    field(DTYP, "stream")
+    field(OUT,  "@mk2chopper.proto setChopperStarted $(PORT)")
+    field(EGU, "")
```
Comment (Tom-Willemsen) [inline MK2CHOPR/MK2CHOPR-IOC-01App/Db/MK2CHOPR.db]: From my IOC logs:

```
[...] Record "TE:NDW1298:MK2CHOPR_01:ON:SP" does not have a field "EGU"
[...] Error at or before ")" in file "C:/Instrument/Apps/EPICS/ioc/master/MK2CHOPR/db/MK2CHOPR.db" line 197
```

There's another very similar issue on line 214.

File: MK2CHOPR/MK2CHOPR-IOC-01App/Db/MK2CHOPR.db [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+  field(SIML, "$(P)SIM")
+  field(SIOL, "$(P)SIM:ERR:STAT")
+}
+
+record(mbbiDirect, "$(P)SIM:ERR:STAT")
+{
+}
+
+record(bi, "$(P)ERR:PHASE:DELAY") {
+  field(DESC, "The phase delay is wrong for the current speed")
```
Comment (Tom-Willemsen) [inline MK2CHOPR/MK2CHOPR-IOC-01App/Db/MK2CHOPR.db]: from IOC logs

```
[...] Can't set "TE:NDW1298:MK2CHOPR_01:ERR:PHASE:DELAY.DESC" to "The phase delay is wrong for the current speed"
[...] Error at or before ")" in file "C:/Instrument/Apps/EPICS/ioc/master/MK2CHOPR/db/MK2CHOPR.db" line 367
```

Similar problem on line 385.

File: MK2CHOPR/MK2CHOPR-IOC-01App/Db/MK2CHOPR.db [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+  field(SCAN, "1 second")  
+  field(DTYP, "stream")
+  field(INP, "@mk2chopper.proto getChopperLocks $(PORT)")
+  field(SIML, "$(P)SIM")
+  field(SIOL, "$(P)SIM:CHOP:STAT")
+}
+
+record(mbbiDirect, "$(P)SIM:CHOP:STAT")
+{
+}
```
Comment (Tom-Willemsen) [inline MK2CHOPR/MK2CHOPR-IOC-01App/Db/MK2CHOPR.db]: Should this record have a `DESC` field at least (if it doesn't need anything else?)

File: MK2CHOPR/MK2CHOPR-IOC-01App/Db/MK2CHOPR.db [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+  field(SIML, "$(P)SIM")
+  field(SIOL, "$(P)SIM:SPEC:STAT")
+}
+
+record(mbbiDirect, "$(P)SIM:SPEC:STAT")
+{
+}
+
+record(bi, "$(P)SPEC:STAT:DRIVE") {
+  field(DESC, "Wheter the drive is running")
```
Comment (Tom-Willemsen) [inline MK2CHOPR/MK2CHOPR-IOC-01App/Db/MK2CHOPR.db]: Whether 

:-)

File: Makefile [hunk 5]
```diff
@@ -8,7 +8,7 @@ BUILDING_SHARED=YES
 endif
 
 ## list all valid IOC directories that we may want to build at some point
-IOCDIRS = AG33220A AG3631A AG53220A CCD100 CONEXAGP CONTROLSVCS CRYVALVE DELFTARDUSTEP DELFTBPMAG DELFTDCMAG DELFTSHEAR DFKPS ECLAB EUROTHRM FINS GALIL HAMEG8123 HIFIMAG HVCAEN INSTETC ISISDAE JULABO KHLY2400 KEPCO LINKAM95 LKSH218 LKSH336 MCLEN MERCURY_ITC MK3CHOPR NANODAC NEOCERA PDR2000 PIMOT PSCTRL RUNCTRL SDTEST SKFCHOPPER SMC100 SPINFLIPPER306015 STPS350 STSR400 TDK_LAMBDA_GENESYS TEKAFG3XXX TEKDMM40X0 TEKMSO4104B TEST TPG26x TPG300 TTIEX355P SCIMAG3D
+IOCDIRS = AG33220A AG3631A AG53220A CCD100 CONEXAGP CONTROLSVCS CRYVALVE DELFTARDUSTEP DELFTBPMAG DELFTDCMAG DELFTSHEAR DFKPS ECLAB EUROTHRM FINS GALIL HAMEG8123 HIFIMAG HVCAEN INSTETC ISISDAE JULABO KHLY2400 KEPCO LINKAM95 LKSH218 LKSH336 MCLEN MERCURY_ITC MK3CHOPR MK2CHOPR NANODAC NEOCERA PDR2000 PIMOT PSCTRL RUNCTRL SCIMAG3D SDTEST SKFCHOPPER SMC100 SPINFLIPPER306015 STPS350 STSR400 TDK_LAMBDA_GENESYS TEKAFG3XXX TEKDMM40X0 TEKMSO4104B TEST TPG26x TPG300 TTIEX355P
```
Comment (Tom-Willemsen) [inline Makefile]: Nit-picking, but if we're doing alphabetical order then "MK2" should come before "MK3"

File: MK2CHOPR/MK2CHOPR-IOC-01App/Db/MK2CHOPR.db (hunk 1 above)
Comment (AdrianPotter) [inline MK2CHOPR/MK2CHOPR-IOC-01App/Db/MK2CHOPR.db]: Huh, thought I removed those. Thanks

File: MK2CHOPR/MK2CHOPR-IOC-01App/Db/MK2CHOPR.db (hunk 3 above)
Comment (AdrianPotter) [inline MK2CHOPR/MK2CHOPR-IOC-01App/Db/MK2CHOPR.db]: We don't typically give description fields for SIM records. I'll double check what the standards say

File: MK2CHOPR/MK2CHOPR-IOC-01App/Db/MK2CHOPR.db (hunk 4 above)
Comment (AdrianPotter) [inline MK2CHOPR/MK2CHOPR-IOC-01App/Db/MK2CHOPR.db]: Fair point :)

File: MK2CHOPR/MK2CHOPR-IOC-01App/Db/MK2CHOPR.db (hunk 2 above)
Comment (AdrianPotter) [inline MK2CHOPR/MK2CHOPR-IOC-01App/Db/MK2CHOPR.db]: Will look into this. Those types of errors are often a red herring for something else

File: Makefile (hunk 5 above)
Comment (AdrianPotter) [inline Makefile]: image

File: MK2CHOPR/MK2CHOPR-IOC-01App/Db/MK2CHOPR.db (hunk 3 above)
Comment (AdrianPotter) [inline MK2CHOPR/MK2CHOPR-IOC-01App/Db/MK2CHOPR.db]: As per https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Record-Simulation I don't see descriptions as necessary on sim records.

File: MK2CHOPR/MK2CHOPR-IOC-01App/Db/MK2CHOPR.db (hunk 2 above)
Comment (AdrianPotter) [inline MK2CHOPR/MK2CHOPR-IOC-01App/Db/MK2CHOPR.db]: Nothing nefarious. The description was just too long for EPICS

**Pull request merged**